src/imageio/format/jpeg.c

```diff
@@ -358,23 +395,32 @@ void *legacy_params(dt_imageio_module_format_t *self,
     return n;
   }
 
-  // incremental update supported:
-  /*
   typedef struct dt_imageio_jpeg_v3_t
   {
-    ...
+    dt_imageio_module_data_t global;
+    int quality;
+    struct jpeg_source_mgr src;
+    struct jpeg_destination_mgr dest;
+    struct jpeg_decompress_struct dinfo;
+    struct jpeg_compress_struct cinfo;
+    FILE *f;
+    int subsample;
   } dt_imageio_jpeg_v3_t;
 
   if(old_version == 2)
   {
-    // let's update from 2 to 3
+    // Version 3 only adds subsample field at the end of the struct
+    // so we use memcpy for simplicity
+    const dt_imageio_jpeg_v2_t *o = (dt_imageio_jpeg_v2_t *)old_params;
+    dt_imageio_jpeg_v3_t *n = (dt_imageio_jpeg_v3_t *)malloc(sizeof(dt_imageio_jpeg_v3_t));
+    memcpy(n, o, sizeof(struct dt_imageio_jpeg_v2_t));
+    n->subsample = 0; // auto
 
-    ...
-    *new_size = sizeof(dt_imageio_module_data_t) + sizeof(int);
     *new_version = 3;
+    *new_size = sizeof(dt_imageio_module_data_t) + sizeof(int);
```

Comment:
I understand that this is what I have put in the comment when working on incremental legacy_params but it looks wrong and probably a copy/paste from the handling of migration from v1 to v2.

You have added an int (should be after quality) and we want `+ 2 * sizeof(int)` for this v3.

That is, we store all data up to `struct jpeg_source_mgr src;` which is only for the internal stuff.

Looks at other format, this is tricky.

Reply:
> I understand that this is what I have put in the comment when working on incremental legacy_params but it looks wrong and probably a copy/paste from the handling of migration from v1 to v2.

Am I correct in understanding that the v1 to v2 migration code I used as an example also contains a bug? Looking at other formats, I noticed that `sizeof(dt_imageio_module_data_t)` definitely shouldn't be here.

Migration from v1 to v2 contains the following code:

```*new_size = sizeof(dt_imageio_module_data_t) + sizeof(int);```

but it should, as I understand, be

```*new_size = sizeof(dt_imageio_jpeg_v2_t) + sizeof(int);```

Also, why do we do summation when it's logical to just take the size of the new structure, like, for v2->v3:

```*new_size = sizeof(dt_imageio_jpeg_v3_t);```

Is this correct or am I still missing something about the migration mechanism?

Reply:
> *new_size = sizeof(dt_imageio_module_data_t) + sizeof(int);

This is correct.

> *new_size = sizeof(dt_imageio_jpeg_v2_t) + sizeof(int);

This is not.

The definition in v2 is:

```
  typedef struct dt_imageio_jpeg_v2_t
  {
    dt_imageio_module_data_t global;
    int quality;
    struct jpeg_source_mgr src;
    struct jpeg_destination_mgr dest;
    struct jpeg_decompress_struct dinfo;
    struct jpeg_compress_struct cinfo;
    FILE *f;
  } dt_imageio_jpeg_v2_t;
```

The size of the data is what we want to serialize in `get_params` & `set_params`.

The actual data to store are : global & quality. The other fields are for the internal use hence the:

`*new_size = sizeof(dt_imageio_module_data_t) + sizeof(int);`

Now for v2 you have a new int (well enum) and so data to serialize are: `global` & `quality` & `subsample`, so:

`*new_size = sizeof(dt_imageio_module_data_t) + 2 * sizeof(int);`

That is also why `subsample` field must be put just after `quality`.

Hope this clarifies the issue.

Reply:
Or clarity for v3:

`*new_size = sizeof(dt_imageio_module_data_t) + sizeof(int) + sizeof(dt_imageio_jpeg_subsample_t);`

Reply:
@victoryforce : A side note, you may understand now why I wanted this into 4.8, all those reviews needs time and energy. And that's not a criticism about your work, just that some PR are really demanding and I sometimes I fell short on time.

Reply:
> The size of the data is what we want to serialize in get_params & set_params.


Ah, now it is clear. Thanks. It was not entirely obvious, and the code for other formats did not have such data "for internal use", that we should exclude, so it did not work out to take them as a sample.

> A side note, you may understand now why I wanted this into 4.8

I perfectly understand this, as well as the stress you have to withstand before freezing the code and how much energy you dedicate to reviews, I just think that the value of the functionality to the user in certain cases may be more valuable than the strict adherence to the planned moment of freezing the code. After all, it doesn't even push the planned release.

Reply:
> and the code for other formats did not have such data "for internal use"

The TIFF format as something similar:

`    *new_size = sizeof(dt_imageio_tiff_v4_t) - sizeof(TIFF *);`